File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Java/11-申请交易账单/GetTradeBill.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/partner/4014985777

import com.google.gson.annotations.SerializedName;
import com.google.gson.annotations.Expose;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 申请交易账单API
 */
public class GetTradeBill {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "GET";
  private static String PATH = "/v3/bill/tradebill";

  public static void main(String[] args) {
    // TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/partner/4013080340
    GetTradeBill client = new GetTradeBill(
      "YOUR_MCHID",                    // 商户号，是由微信支付系统生成并分配给每个商户的唯一标识符，商户号获取方式参考 https://pay.weixin.qq.com/doc/v3/partner/4013080340
      "YOUR_CERT_SERIAL_NO",         // 商户API证书序列号，如何获取请参考 https://pay.weixin.qq.com/doc/v3/partner/4013058924
      "/path/to/apiclient_key.pem",     // 商户API证书私钥文件路径，本地文件路径
      "YOUR_PUB_KEY_ID",      // 微信支付公钥ID，如何获取请参考 https://pay.weixin.qq.com/doc/v3/partner/4013038589
      "/path/to/wxp_pub.pem"           // 微信支付公钥文件路径，本地文件路径
    );

    GetTradeBillRequest request = new GetTradeBillRequest();
    request.billDate = "2019-06-11";
    request.subMchid = "19000000001";
    request.billType = BillType.ALL;
    request.tarType = TarType.GZIP;
    try {
      QueryBillEntity response = client.run(request);
        // TODO: 请求成功，继续业务逻辑
        System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
        // TODO: 请求失败，根据状态码执行不同的逻辑
        e.printStackTrace();
    }
  }

  public QueryBillEntity run(GetTradeBillRequest request) {
    String uri = PATH;
    Map<String, Object> args = new HashMap<>();
    args.put("bill_date", request.billDate);
    args.put("sub_mchid", request.subMchid);
    args.put("bill_type", request.billType);
    args.put("tar_type", request.tarType);
    uri = uri + "?" + WXPayUtility.urlEncode(args);

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo, privateKey, METHOD, uri, null));
    reqBuilder.method(METHOD, null);
    Request httpRequest = reqBuilder.build();

    // 发送HTTP请求
    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        // 2XX 成功，验证应答签名
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);

        // 从HTTP应答报文构建返回数据
        return WXPayUtility.fromJson(respBody, QueryBillEntity.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public GetTradeBill(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class GetTradeBillRequest {
    @SerializedName("bill_date")
    @Expose(serialize = false)
    public String billDate;

    @SerializedName("sub_mchid")
    @Expose(serialize = false)
    public String subMchid;

    @SerializedName("bill_type")
    @Expose(serialize = false)
    public BillType billType;

    @SerializedName("tar_type")
    @Expose(serialize = false)
    public TarType tarType;
  }

  public static class QueryBillEntity {
    @SerializedName("hash_type")
    public HashType hashType;

    @SerializedName("hash_value")
    public String hashValue;

    @SerializedName("download_url")
    public String downloadUrl;
  }

  public enum BillType {
    @SerializedName("ALL")
    ALL,
    @SerializedName("SUCCESS")
    SUCCESS,
    @SerializedName("REFUND")
    REFUND
  }

  public enum TarType {
    @SerializedName("GZIP")
    GZIP
  }

  public enum HashType {
    @SerializedName("SHA1")
    SHA1
  }

}

File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Go/11-申请交易账单/GetTradeBill.go
package main

import (
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/partner/4015119446
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
)

func main() {
	// TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/partner/4013080340
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",                 // 商户号，是由微信支付系统生成并分配给每个商户的唯一标识符，商户号获取方式参考 https://pay.weixin.qq.com/doc/v3/partner/4013080340
		"YOUR_CERT_SERIAL_NO",      // 商户API证书序列号，如何获取请参考 https://pay.weixin.qq.com/doc/v3/partner/4013058924
		"/path/to/apiclient_key.pem", // 商户API证书私钥文件路径，本地文件路径
		"YOUR_PUB_KEY_ID",   // 微信支付公钥ID，如何获取请参考 https://pay.weixin.qq.com/doc/v3/partner/4013038589
		"/path/to/wxp_pub.pem",       // 微信支付公钥文件路径，本地文件路径
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &GetTradeBillRequest{
		BillDate: wxpay_utility.String("2019-06-11"),
		SubMchid: wxpay_utility.String("19000000001"),
		BillType: BILLTYPE_ALL.Ptr(),
		TarType:  TARTYPE_GZIP.Ptr(),
	}

	response, err := GetTradeBill(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		// TODO: 请求失败，根据状态码执行不同的处理
		return
	}

	// TODO: 请求成功，继续业务逻辑
	fmt.Printf("请求成功: %+v\n", response)
}

func GetTradeBill(config *wxpay_utility.MchConfig, request *GetTradeBillRequest) (response *QueryBillEntity, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "GET"
		path   = "/v3/bill/tradebill"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	query := reqUrl.Query()
	query.Add("bill_date", *request.BillDate)
	query.Add("sub_mchid", *request.SubMchid)
	query.Add("bill_type", fmt.Sprintf("%v", *request.BillType))
	query.Add("tar_type", fmt.Sprintf("%v", *request.TarType))
	reqUrl.RawQuery = query.Encode()
	httpRequest, err := http.NewRequest(method, reqUrl.String(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		// 2XX 成功，验证应答签名
		err = wxpay_utility.ValidateResponse(
			config.WechatPayPublicKeyId(),
			config.WechatPayPublicKey(),
			&httpResponse.Header,
			respBody,
		)
		if err != nil {
			return nil, err
		}
		response := &QueryBillEntity{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}

		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(
			httpResponse.StatusCode,
			httpResponse.Header,
			respBody,
		)
	}
}

type GetTradeBillRequest struct {
	BillDate *string   `json:"bill_date,omitempty"`
	SubMchid *string   `json:"sub_mchid,omitempty"`
	BillType *BillType `json:"bill_type,omitempty"`
	TarType  *TarType  `json:"tar_type,omitempty"`
}

func (o *GetTradeBillRequest) MarshalJSON() ([]byte, error) {
	type Alias GetTradeBillRequest
	a := &struct {
		BillDate *string   `json:"bill_date,omitempty"`
		SubMchid *string   `json:"sub_mchid,omitempty"`
		BillType *BillType `json:"bill_type,omitempty"`
		TarType  *TarType  `json:"tar_type,omitempty"`
		*Alias
	}{
		BillDate: nil,
		SubMchid: nil,
		BillType: nil,
		TarType:  nil,
		Alias:    (*Alias)(o),
	}
	return json.Marshal(a)
}

type QueryBillEntity struct {
	HashType    *HashType `json:"hash_type,omitempty"`
	HashValue   *string   `json:"hash_value,omitempty"`
	DownloadUrl *string   `json:"download_url,omitempty"`
}

type BillType string

func (e BillType) Ptr() *BillType {
	return &e
}

const (
	BILLTYPE_ALL     BillType = "ALL"
	BILLTYPE_SUCCESS BillType = "SUCCESS"
	BILLTYPE_REFUND  BillType = "REFUND"
)

type TarType string

func (e TarType) Ptr() *TarType {
	return &e
}

const (
	TARTYPE_GZIP TarType = "GZIP"
)

type HashType string

func (e HashType) Ptr() *HashType {
	return &e
}

const (
	HASHTYPE_SHA1 HashType = "SHA1"
)
